store/cma_store.go

```diff
@@ -0,0 +1,135 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+// cmaStore maintains the cumulative moving average of values at a specific Timestamp.
+// cmaStore is an implementation of TimeStore where the values are casted as uint64.
+type cmaStore struct {
+	buffer *list.List
+	rwLock sync.RWMutex
+}
+
+type TPContainer struct {
+	TimePoint
+	count uint16
+}
+
+func (ts *cmaStore) Put(tp TimePoint) error {
+	if tp.Value == nil {
+		return fmt.Errorf("cannot store TimePoint with nil data")
+	}
+	if (tp.Timestamp == time.Time{}) {
+		return fmt.Errorf("cannot store TimePoint with zero timestamp")
+	}
+	ts.rwLock.Lock()
+	defer ts.rwLock.Unlock()
+	new_tpc := TPContainer{
+		TimePoint: tp,
+		count:     uint16(1),
+	}
+	if ts.buffer.Len() == 0 {
+		glog.V(5).Infof("put pushfront: %v, %v", tp.Timestamp, tp.Value)
+		ts.buffer.PushFront(new_tpc)
+		return nil
+	}
+	for elem := ts.buffer.Front(); elem != nil; elem = elem.Next() {
+		tpc := elem.Value.(TPContainer)
+		if tp.Timestamp.Equal(tpc.Timestamp) {
+			// If an element with that timestamp exists, update its average and count
+			new_val := tp.Value.(uint64)
+			n := uint64(tpc.count)
+			old_avg := tpc.Value.(uint64)
+			tpc.Value = uint64((new_val + (n * old_avg)) / (n + 1))
+			tpc.count = tpc.count + 1
+			elem.Value = tpc
+			return nil
+		} else if tp.Timestamp.After(tpc.Timestamp) {
+			glog.V(5).Infof("put insert before: %v, %v, %v", elem, tp.Timestamp, tp.Value)
+			ts.buffer.InsertBefore(new_tpc, elem)
+			return nil
+		}
+	}
+	glog.V(5).Infof("put pushback: %v, %v", tp.Timestamp, tp.Value)
+	ts.buffer.PushBack(new_tpc)
+	return nil
+}
+
+func (ts *cmaStore) Get(start, end time.Time) []TimePoint {
+	zeroTime := time.Time{}
+	ts.rwLock.RLock()
+	defer ts.rwLock.RUnlock()
+	if ts.buffer.Len() == 0 {
+		return nil
+	}
+	result := []TimePoint{}
+	for elem := ts.buffer.Front(); elem != nil; elem = elem.Next() {
+		tpc := elem.Value.(TPContainer)
+		entry := tpc.TimePoint
+
+		// Skip entries until the first one after start
+		if (start != zeroTime) && entry.Timestamp.Before(start) {
+			continue
+		}
+		// Add all entries whose timestamp is not after end.
+		if (end == time.Time{}) || !entry.Timestamp.After(end) {
+			result = append(result, entry)
+		}
+	}
+	return result
+}
+
+func (ts *cmaStore) Delete(start, end time.Time) error {
```

Comment:
in_memory.go is used only for the cache implementation, it is barely documented and it is not as well-tested as cma_store. As the cache might (or might not) be eventually deprecated, in_memory might be removed as well, or just replaced by cma_store. Duplication was used now to avoid taking such deprecation decisions. 

I expect that the functionality of in_memory.go, if not deprecated, can be implemented within cma_store with a simple configuration flag and a conditional statement.